### ISSUE TITLE
Inconsistent error handling and documentation in COPYPATH

### ISSUE THREAD

**Issue by bfrobin446** [KSP-KOS/KOS#1740]
When COPYPATH fails because the source file is not found, an exception is thrown with a "Path not found" message. If it fails because there's not enough free space on the destination volume, there's no exception. It just returns `false`. I can't find anything in the documentation that says these functions have a meaningful return value, so I spent fifteen or twenty minutes trying to figure out why my files weren't copying before it occurred to me to check the return value.

* Is there a design reason why COPYPATH reports different errors in different ways?

* Once there's a consensus on how the file system functions should report errors, I'd be willing to take a pass through the documentation and make sure that for any operation that can fail, we're documenting how it reports that failure.

**Comment by hvacengi** [comment]
I'm gonna ping @tomekpiotrowski for comment here, since he did much of the work surrounding support of paths.

**Comment by Dunbaratu** [comment]
Hmm.  The problem that "not found" throws an error, while "volume full" just silently fails without complaint was true long before @tomekpiotrowski implemented his changes.  He probably just inherited that inconsistency because he was trying to keep things as similar as possible to the old way, and built on top of the way it worked before.

I have no clue why it was that way originally, though.  We really should pick one and stick with it.

I'd normally favor throwing an exception, were it not for the fact that kerboscript doesn't do exception handling so it would make the script writer powerless to detect it and do something about it.

**Comment by hvacengi** [comment]
Yeah, if we want to throw an error, we really need a `hasSpaceFor` suffix of some sort.

**Comment by Dunbaratu** [comment]
I seriously would like to allow exception handling some day.  The ugly part is not how to implement it - that's straightforward - we walk the call stack global-ward until we find the first trapped section with the exception mentioned in its whitelist of what to trap.  The ugly part is how to make sure we don't leave our *own* virtual machine in C# in a wonky broken state.  We'd basically have to look at all the places we throw KOSExceptions (and its derivatives), and make sure we're not leaving valuable important housekeeping work un-executed due to those early aborts.

But that's a topic for another time.  The ``hasSpaceFor`` suffix would suffice under the current "check before you try using it" model that people have to use for error handling.

**Comment by tomekpiotrowski** [comment]
For me actually the distinction here is clear. I've tried to use exceptions for things that constitute a programmer error. In this case if you're supplying an invalid path as an argument it is your mistake. A return value represents the result of a method call when it was called correctly.

Btw, I'm all for a `:HASSPACEFOR` suffix :).

**Comment by Dunbaratu** [comment]
> if you're supplying an invalid path as an argument it is your mistake

If we ever get user input working that won't be true anymore.
```
Welcome to HAL900, Dave.
Which program are you trying to run> asdfasdfasdfasdf
Sorry, I could not find 'asdfasdfasdf', Dave.
```
It's not the programmer of HAL's fault that Dave typed in garbage.
(Or imagine being the writer of `bash` and having to deal with a user typing in `rm alksdjfalksdjlfk` without having `bash` crash because of it.)

**Comment by KSAMissionCtrl** [comment]
This is still an issue that needs to be resolved - I just spent a good 15-20 minutes trying to figure out why a file wasn't copying when using the copypath command for the first time since upgrading to this newer version with directory structure. Finally I noticed the filesize was too big for the storage capacity of my kOS unit.

*also* I'm not sure if this warrants its own issue but the file I was trying to copy was the boot file because kOS was logging an error to the console that the file could not be found when the real issue was the file was too big - so that error message sent me off on a wild goose chase looking at my folder structure

**Comment by Dunbaratu** [comment]
The error message that it cannot find the file is correct and shouldn't be "fixed". 

The problem is caused by the earlier step - when copying the file from the archive *to* the local volume - not issuing a complaint message when it doesn't fit and therefore fails to make the local copy.

Once the local copy doesn't exist, it's correct to say it can't find it.  (Note the filename it's complaining about is "/boot/whatever" rather than "0:/boot/watever".  It's not saying it can't find it on the archive volume.  It's saying it cant' find it on the local volume.  Which is true.  Because it didn't get copied.)

The problem should be fixed, you are right, but the fix isn't to change that message.  That message is right.  The fix is to also insert a message on the prior step, so that the reason the file is missing is made clear.

There's actually two things to do.  One is to raise the exception.  The second is to make special case code in the boot file initialization process that will communicate this message to the user in an unusual way.  (The /boot file copy that occurs during vessel launch for the boot file happens outside the context of the terminal interpreter, so it wouldn't normally appear on screen like all other exceptions would.)

**Comment by Dunbaratu** [comment]
It may also be good to declare the message in the VAB/SPH when the boot file is selected in the first place.
If you pick a boot file that's bigger than the kOS local volume, it could issue a complaint then and there, before you pay Funds to launch the ship to the launchpad.

**Comment by hvacengi** [comment]
Maybe we can add a readout for "free space" that turns red or something when it goes negative. Or a pop up message before allowing the launch.

**Comment by KSAMissionCtrl** [comment]
I had no idea the "file not found" message was telling me that it wasn't finding the file on the local drive. I thought it was telling me it couldn't find the file on the archive drive. Since I knew it had to be copied before it could be run, like you said I was viewing the error in context of the first step. It may be beneficial to make clarify that that error is a run error not a copy error

**Comment by Tabris83** [comment]
Just chiming in that this issue is still occuring.

**Comment by Tahvohck** [comment]
Bumping this again. Just spent a while trying to figure out why copyPath wasn't working. It's only because I'm a technical user that I ended up here and found out that my file was too big.
